File: LeetCode/Medium/0622-design-circular-queue/0622-design-circular-queue.java
class MyCircularQueue {
    int[] q;
    int front, rear, count;

    public MyCircularQueue(int k) {
        this.q = new int[k];
        this.front = 0;
        this.rear = -1;
        this.count = 0;
    }

    public boolean enQueue(int value) {
        if (isFull()) return false;

        rear = (rear + 1) % q.length;
        q[rear] = value;
        count++;
        return true;
    }

    public boolean deQueue() {
        if (isEmpty()) return false;

        front = (front + 1) % q.length;
        count--;
        return true;
    }

    public int Front() {
        return (isEmpty() ? -1 : q[front]);
    }

    public int Rear() {
        return (isEmpty() ? -1 : q[rear]);
    }

    public boolean isEmpty() {
        return count == 0;
    }

    public boolean isFull() {
        return count == q.length;
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */

File: LeetCode/Medium/0622-design-circular-queue/0622-design-circular-queue.kt
class MyCircularQueue(k: Int) {
    val q = IntArray(k)
    var front = 0
    var rear = -1
    var count = 0

    fun enQueue(value: Int): Boolean {
        if (this.isFull()) return false

        this.rear = (this.rear + 1) % this.q.size
        this.q[rear] = value
        this.count++
        return true
    }

    fun deQueue(): Boolean {
        if (this.isEmpty()) return false

        this.front = (this.front + 1) % this.q.size
        this.count--
        return true;
    }

    fun Front(): Int = if (this.isEmpty()) -1 else this.q[this.front]

    fun Rear(): Int = if (this.isEmpty()) -1 else this.q[this.rear]

    fun isEmpty(): Boolean = this.count == 0

    fun isFull(): Boolean = this.count == this.q.size
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * var obj = MyCircularQueue(k)
 * var param_1 = obj.enQueue(value)
 * var param_2 = obj.deQueue()
 * var param_3 = obj.Front()
 * var param_4 = obj.Rear()
 * var param_5 = obj.isEmpty()
 * var param_6 = obj.isFull()
 */